### PR TITLE
feat(rule): add version compatibility rules

### PR DESCRIPTION
**Pull request by Oleksiivanov**
related : 
 - https://github.com/camunda/team-connectors/issues/733

## Changes Include:
- Added new version compatibility rule for Camunda Cloud versions 8.0 + 
- Included tests to validate rules across different versions.

## Example:

Imagine there's a new feature in version 8.4 that isn't in version 8.3. If a user gets a new connector template with a property that version 8.3 doesn't support, they can still use it. But, they will see a warning that tells them about the version issue:

 `The property '${propertyName}' is incompatible with the current camunda runtime version. Please upgrade to version ${version} or later to use this feature.`,

How to add a new property to the rule : 
```javascript
const propertyVersionCompatibility = [
  {
    minimalVersion: '8.6',
    extensionElement: 'zeebe:Properties',
    subExtensionElement: 'properties',
    key: 'name',
    propertyNames: ['messageTtl', 'consumeUnmatchedEvents', 'deduplicationModeManualFlag']
  },
  // Example for outbound connectors:
  {
    minimalVersion: '8.6',
    extensionElement: 'zeebe:IoMapping',
    subExtensionElement: 'inputParameters',
    key: 'target',
    propertyNames: ['newPropertyName']
  }
];

### CODE REVIEW

**Comment by nikku** [comment]
@Oleksiivanov Could you elaborate a little more on how this works and how you ensure that you do not interfere with other non-connector use-cases?

I see that we match for certain properties (`zeebe:property` but also `zeebe:ioMapping`). These can be used by _arbitary_ tasks, for arbitrary use-cases. The least thing we want to do is to ensure that a certain `zeebe:topic` (task worker) is configured I think, before we go on to validate generic properties that anyone can technically use.

File: rules/camunda-cloud/version-compatibility-validator.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  return {
+    check
+  };
+});
+
+function getReport(propertyName, node, parentNode, version) {
+  const path = getPath(node, parentNode);
+
+  return {
+    message: `The property '${propertyName}' is incompatible with the current camunda runtime version. Please upgrade to version ${version} or later to use this feature.`,
```
[the remark this answers is not in the document]
Comment (nikku) [reply rules/camunda-cloud/version-compatibility-validator.js:71]: @philippfromme Maybe you could have a look and guide @Oleksiivanov regarding other things that are important for version compatibility?

We want to allow users to learn that certain connector properties are only supported in 8.6 (or newer) -> https://github.com/camunda/team-connectors/issues/733.

**Comment by nikku** [comment]
@Oleksiivanov I see two options:

* :a: You could match elements with a specific modeler template assigned
* :b: You could match elements (service tasks) with a specific `zeebe:taskDefinition#type` assigned or another property (for inbound) that identifies the element from the connectors run-time point of view

Either one would accomplish different things:

* :a: stops to work when users customize the template, most strict (least false positives when matching)
* :b: is less strict; allows users to change the element template while still benefiting from the warnings; validates also when element templates are unbound; may cause false positives in the rare case where folks do not use connectors, but re-use the technical bindings that connectors build upon (I honestly think we can ignore this!)

If you ask me for a recommendation, then I'd go with :b:, because it reflects the run-time constraints _independent_ of our OOTB element templates.

File: index.js [hunk 2]
```diff
@@ -37,6 +37,7 @@ const camundaCloud13Rules = withConfig(camundaCloud12Rules, { version: '1.3' });
 
 const camundaCloud80Rules = withConfig({
   ...omit(camundaCloud13Rules, 'no-template'),
+  'version-compatibility-validator': 'warn'
```
Comment (nikku) [inline index.js:40]: We probably want to find a better name for this.

File: rules/camunda-cloud/version-compatibility-validator.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
+const {
+  findExtensionElement
+} = require('../utils/element');
+
+const { reportErrors } = require('../utils/reporter');
+
+const { skipInNonExecutableProcess } = require('../utils/rule');
```
Comment (nikku) [inline rules/camunda-cloud/version-compatibility-validator.js:7]: I'm not sure if this fully utilizes the existing infrastructure we have in place. I asked @philippfromme to review, too.

File: index.js (hunk 2 above)
Comment (nikku) [reply index.js:40]: Waiting for @philippfromme to share his cents.

**Comment by nikku** [comment]
@philippfromme Please review this PR.

**Comment by philippfromme** [comment]
Looking at this and thinking about the long term I'd argue we want

* separate rules for each connector template
* put all these rules in a separate linter plugin

Otherwise I'm worried about the maintainability of this plugin. Also, this plugin is strictly about engine compatibility and you could argue that the connector template rules are about engine plus connectors runtime compatibility.

@nikku What do you think? Overkill?

**Comment by nikku** [comment]
@philippfromme If we keep it here then we have good practices under control.

The least thing we could do is to prefix / namespace connector related rules, so that pulling them out will be possible in the future.

File: index.js (hunk 2 above)
Comment (philippfromme) [reply index.js:40]: This definitely needs to be renamed. The scope of the plugin is bigger than just connector templates so we should namespace connector template specific things.

File: rules/camunda-cloud/version-compatibility-validator.js (hunk 3 above)
Comment (philippfromme) [reply rules/camunda-cloud/version-compatibility-validator.js:7]: 👀

**Comment by philippfromme** [comment]
Okay, let's keep the rule(s) in this plugin but reorganize a bit. I'm looking into this now.

**Comment by philippfromme** [comment]
I made some changes:

* renamed the rule to `connector-properties`
* refactored the implementation
  * split into `index` and `config`
  * report contains all necessary data to [create a user-friendly error message](https://github.com/camunda/linting/blob/main/lib/utils/error-messages.js)
* removed code that isn't used at the moment (no outbound connector properties linted yet)

brave_bLsnAwEBYM

**Comment by nikku** [comment]
Great job both :clap:

**Pull request merged**